doc/debugger.rst

```diff
@@ -1773,6 +1808,74 @@ target endianness and interpret the bytecode based on that.
 .. note:: This command is somewhat incomplete at the moment and may be modified
    once the best way to do this in the debugger UI has been figured out.
 
+AppRequest request (0x23)
+-------------------------
+
+Format::
+
+    REQ <int: 0x23> [<tval>*] EOM
+    REP [<tval>*] EOM
+
+Example::
+
+    REQ 35 "GameInfo" "GetTitle" EOM
+    REP "Spectacles: Bruce's Story" EOM
+
+This is a custom request message whose meaning and semantics depend on the
+application.
+
+AppRequest messages are used for direct communication between the debug client
+and debug target over the Duktape debug protocol. Both the meaning of a custom
+message and the dvalues it contains are entirely up to the implementation and
+depending on the needs of the application, need not be supported at all.
+
+See "Custom requests and notifications" below for more details.
+
+Custom requests and notifications
+=================================
+
+Starting in Duktape 1.5.x, Duktape supports direct communication between the
+debug client and debug target over the same transport by using the special
+'AppRequest' and 'AppNotify' messages. These messages have no meaning to
+Duktape, which merely serves to marshal them back and forth through a defined
+API.
+
+AppNotify messages may be sent by pushing the contents of the message to the
+stack and calling `duk_debugger_notify()` passing the number of values pushed.
+Each value pushed will be sent as a dvalue in the message. So if you push two
+strings, "foo" and "bar", the client will see ``NFY 7 "foo" "bar" EOM``.
+
+AppRequests are used to make requests to the target which are not directly
+related to Ecmascript execution. For example, a game engine might allow the
+debug client to change the framerate, or request information about the
+currently running game. Supporting this requires attaching using the
+``duk_debugger_attach_custom()`` API and providing a request callback. When an
+AppRequest is received, the request callback is invoked with the contents of
+the message on the value stack, and may push its own values to be sent in
+reply.
+
+This is a do-nothing request callback::
+
+    duk_idx_t duk_cb_debug_request(void* udata, duk_context* ctx, duk_idx_t nvalues) {
+	    /* nop */
+	    return 0;
+    }
+
+This simply responds with ``REP EOM`` (an empty reply) to all requests. This is
+not very useful and a real implementation should process the values it receives
+on the stack, push its own values to send in reply, and return a non-negative
+integer indicating how many values it pushed.
```

Comment:
The "do nothing" example is OK to illustrate what a zero return means, but perhaps add an example of returning e.g. a string, a number, and a boolean (for example)?